go-controller/pkg/ovn/default_network_controller.go

```diff
@@ -497,11 +501,21 @@ func (oc *DefaultNetworkController) Run(ctx context.Context) error {
 
 	if config.OVNKubernetesFeature.EnableEgressFirewall {
 		var err error
-		oc.egressFirewallDNS, err = NewEgressDNS(oc.addressSetFactory, oc.controllerName, oc.stopChan)
+		// If DNSNameResolver is enabled, then initialize dnsNameResolver to ExternalEgressDNS
+		// for maintaining the address sets corresponding to the DNS names and start watching
+		// DNSNameResolver resources. Otherwise initialize dnsNameResolver to EgressDNS.
+		if config.OVNKubernetesFeature.EnableDNSNameResolver {
+			oc.dnsNameResolver, err = dnsnameresolver.NewExternalEgressDNS(oc.addressSetFactory, oc.controllerName, true, oc.watchFactory.DNSNameResolverInformer().Informer())
+		} else {
+			oc.dnsNameResolver, err = dnsnameresolver.NewEgressDNS(oc.addressSetFactory, oc.controllerName, oc.stopChan, egressFirewallDNSDefaultDuration)
+		}
+		if err != nil {
+			return err
+		}
+		err = oc.dnsNameResolver.Run()
```

Comment:
One more this I would like to fix before merging: cleanup on restart.
So imagine ovn-k got restarted and in the meantime one egressfirewall (or a rule in egressfirewall) was deleted together with related DNSNameResolver object. We won't receive any delete events on restart. So I have an idea how we can fix this:
1. dnsNameResolver.Run() should happen after WatchEgressFirewall to make sure our caches are populated with fresh references of namespaces to dnsNames.
2. DNSNameResolver Controller needs an initialSync function that will find all address sets in the database, check which ones of them don't have corresponding DNSNameResolver anymore and delete it from the db.
3. We could use some unit tests for this

Reply:
Updated the code to include this scenario.